### PR TITLE
Change to getClass() from instanceof

### PR DESCRIPTION
**Pull request by eljobe**
This prevents our breaking the symmetrical property of euqals through
extension.

### CODE REVIEW

**Comment by bmuschko** [comment]
@eljobe What's the status on this PR?

**Comment by eljobe** [comment]
@bmuschko, I have asked twice on the build-tool slack channel for a reviewer, but nobody has volunteered. It is awaiting a reviewer. Are you interested?

**Comment by bmuschko** [comment]
@eljobe Happy to review. Likely won't happen until the Summit though.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/adapter/ProtocolToModelAdapter.java [hunk 1]
```diff
@@ -275,6 +275,12 @@ private void readObject(java.io.ObjectInputStream in) throws IOException, ClassN
 
         @Override
         public boolean equals(Object obj) {
```
Comment (bmuschko) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/adapter/ProtocolToModelAdapter.java]: We should add a test that verifies the correct behavior.

**Comment by bmuschko** [comment]
@eljobe Can you provide more background information on your statement above? Do we observe incorrect behavior right now? If yes, how does the behavior manifest?

**Comment by eljobe** [comment]
We do NOT observe incorrect behavior. For more background, you can read this summary:
https://groups.google.com/a/gradle.com/d/msg/build-tool-team/CyNegX5XL-I/er1WdxYvAwAJ

What I'm trying to do is prevent us from introducing a common bug when `equals()` is implemented with `instanceof`. Namely, someone will come along and subclass one of these classes, override the `equals()` method and instantly break the symmetrical property of the `equals()` contract.

So, there is no incorrect behavior now, but as a best practice, we should cleanup any `equals()` implementations which are using `instanceof`.  I'll add some [EqualsVerifier](http://jqno.nl/equalsverifier/) tests for the classes which implement `equals()` and make sure they pass.

**Comment by eljobe** [comment]
@bmuschko, I've added `EqualsVerifer` checks which actually helped catch some minor issues with the implementations of equals() on the types I was cleaning up. This is ready for further review.

**Comment by eljobe** [comment]
@bmuschko ping!?

**Comment by bmuschko** [comment]
@eljobe I'll look at it today.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/adapter/ProtocolToModelAdapter.java [hunk 2]
```diff
@@ -262,7 +264,8 @@ private void readObject(java.io.ObjectInputStream in) throws IOException, ClassN
         }
     }
 
-    private static class ViewKey implements Serializable {
+    @VisibleForTesting
```
Comment (bmuschko) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/adapter/ProtocolToModelAdapter.java]: Nice use of annotation. Was wondering why you made that change.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/adapter/ProtocolToModelAdapter.java [hunk 3]
```diff
@@ -758,14 +770,15 @@ public void invoke(MethodInvocation invocation) throws Throwable {
...
+    static class NoOpDecoration implements ViewDecoration, Serializable {
         @Override
         public void collectInvokers(Object sourceObject, Class<?> viewType, List<MethodInvoker> invokers) {
         }
 
         @Override
         public boolean equals(Object obj) {
-            return obj instanceof NoOpDecoration;
+            return this == obj || obj != null && obj.getClass() == getClass();
```
Comment (bmuschko) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/adapter/ProtocolToModelAdapter.java]: :o: Shouldn't the `obj != null && obj.getClass() == getClass()` be wrapped in parenthesis so it is evaluated as one statement?

Comment (eljobe) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/adapter/ProtocolToModelAdapter.java]: No.

See: https://github.com/eljobe/parentheses/blob/master/src/test/groovy/com/technodabble/booleans/OrAndSpec.groovy

Essentially, the way the boolean operators work, the expression is exactly the same with and without the parenthesis.

**Pull request merged**